README.md
# pystep7
A Python3 implementation of Siemens Step7 protocol

## Installation 
```console 
pip3 install pystep7
```


## How to use pystep7 
### 1. Connect and Send Commands
```python
from datetime import datetime
from pystep7 import Client
from pystep7.constants import Area, BlockType, DataType
from pystep7.tag import Tag, IecCounter, IecTimer


__RTAGS1 = [
    Tag(name="CHAR", address="DB2.DBX2.0", type=DataType.CHAR),
    Tag(name="BYTE", address="DB2.DBX1.0", type=DataType.BYTE),
    Tag(name="DINT", address="DB2.DBX4.0", type=DataType.DINT),
    Tag(name="DWORD", address="DB2.DBX8.0", type=DataType.DWORD),
    Tag(name="S7 DATE", address="DB2.DBX12.0", type=DataType.DATE),
    Tag(name="S7 DATETIME", address="DB2.DBX14.0", type=DataType.DATETIME),
    Tag(name="INT", address="DB2.DBX22.0", type=DataType.INT),
    Tag(name="REAL", address="DB2.DBX24.0", type=DataType.REAL),
    Tag(name="S5 TIME", address="DB2.DBX28.0", type=DataType.S5TIME),
    Tag(name="STRING", address="DB2.DBX30.0", type=DataType.STRING),
    Tag(name="S7 TIME", address="DB2.DBX286", type=DataType.TIME),
    Tag(name="S7 TOD", address="DB2.DBX290", type=DataType.TIME_OF_DAY),
    Tag(name="WORD", address="DB2.DBX294", type=DataType.WORD),
    Tag(name="BOOL0", address="DB2.DBX0.0", type=DataType.BIT),
    Tag(name="BOOL1", address="DB2.DBX0.1", type=DataType.BIT),
    Tag(name="BOOL2", address="DB2.DBX0.2", type=DataType.BIT),
    Tag(name="BOOL3", address="DB2.DBX0.3", type=DataType.BIT),
    Tag(name="BOOL4", address="DB2.DBX0.4", type=DataType.BIT),
    Tag(name="FLAG5", address="M0.4", type=DataType.BIT),
    Tag(name="INPUT3", address="I0.2", type=DataType.BIT),
    Tag(name="COUNTER0", address="C0", type=DataType.COUNTER),  # S7-200
    Tag(name="TIMER0", address="T0", type=DataType.TIMER),      # S7-200
    Tag(name="IEC_COUNTER0", address="T0", type=DataType.IECCOUNTER),   # S7-300+
    Tag(name="IEC_TIMER0", address="T0", type=DataType.IECTIMER),       # S7-300+
]

__WTAGS2 = [
    Tag(name="BOOL0", address="DB2.DBX0.0", value=True, type=DataType.BIT),
    Tag(name="BOOL1", address="DB2.DBX0.1", value=False, type=DataType.BIT),
    Tag(name="BOOL2", address="DB2.DBX0.3", value=True, type=DataType.BIT),
    Tag(name="BOOL3", address="DB2.DBX0.4", value=False, type=DataType.BIT),
    Tag(name="BOOL4", address="DB2.DBX0.5", value=True, type=DataType.BIT),
    Tag(name="BYTE", address="DB2.DBX1.0", value=1, type=DataType.BYTE),
    Tag(name="CHAR", address="DB2.DBX2.0", value='T', type=DataType.CHAR),
    Tag(name="DINT", address="DB2.DBX4.0", value=-10000, type=DataType.DINT),
    Tag(name="DWORD", address="DB2.DBX8.0", value=10000, type=DataType.DWORD),
    Tag(
        name="S7 DATE", 
        address="DB2.DBX12.0",
        value=date(2000,1,1), 
        type=DataType.DATE
    ),
    Tag(
        name="S7 DATETIME", 
        address="DB2.DBX14.0",
        value=datetime(2022,9,4,10,11,12,130000),
        type=DataType.DATETIME
    ),
    Tag(name="INT", address="DB2.DBX22.0", value=-100, type=DataType.INT),
    Tag(name="REAL", address="DB2.DBX24.0", value=6.6, type=DataType.REAL),
    Tag(
        name="S5 TIME",
        address="DB2.DBX28.0",
        value=16990, 
        type=DataType.S5TIME
    ),  # S5 TIME (milliseconds) smallest is 10, largest is 9990000 (2H:46M:30S)
    Tag(
        name="STRING", 
        address="DB2.DBX30.0",
        value='Hello World', 
        type=DataType.STRING
    ),  # STRING (max char is 254)
    Tag(
        name="S7 TIME", 
        address="DB2.DBX286",
        value=1,    # time in milliseconds 
        type=DataType.TIME
    ),  # S7 TIME
    Tag(
        name="S7 TOD", 
        address="DB2.DBX290",
        value=18000000,
        type=DataType.TIME_OF_DAY
    ),  # S7 TIME OF DAY: milliseconds since midnight
    Tag(name="INPUT 3", address="I0.2", value=True, type=DataType.BIT),
    Tag(name="OUTPUT 3", address="Q0.2", value=True, type=DataType.BIT),
    Tag(name="FLAG 5", address="M0.4", value=True, type=DataType.BIT),
    Tag(name="COUNTER0", address="C0", value=100, type=DataType.COUNTER),   # S7-200
    Tag(name="TIMER0", address="T0", value=100, type=DataType.TIMER),       # S7-200 (WIP)
    Tag(
        name="IEC_COUNTER0", 
        address="DB4DBX0.0", 
        value=IecCounter(
            C_DU=True, 
            LOAD_R=True, 
            PV=1000,
            Q=False,
            CV=1000,
            C_DU_O=False
        )
        type=DataType.IECCOUNTER
    ),  # S7-300+
    Tag(
        name="IEC_TIMER", 
        address="DB5DBX0.0", 
        value=IecTimer(
            IN=True,
            PT=1000,    # milliseconds
            Q=True,
            ET=1000,    # milliseconds
            STATE=1,    # milliseconds
            STIME=2,    # milliseconds
            ATIME=3     # milliseconds
        ),
        type=DataType.IECTIMER
    ),  # S7-300+
]
    

__HOST = '192.168.1.15' # REQUIRED
__PORT = 102            # OPTIONAL: default is 102
__RACK = 0              # OPTIONAL: default is 0
__SLOT = 0              # OPTIONAL: default is 0


with Client(host=__HOST, port=__PORT, rack=__RACK, slot=__SLOT) as plc:
    """
    Read PLC time

    Returns:
        datetime: datetime object that represents PLC time
            example:
                2022-09-08 17:07:25.380000
    """
    response = plc.read_plc_time()



    """
    Set PLC time
        example: Set PLC time using local time

    Args:
        dt(datetime): datetime object

    Returns:
        dt(datetime): the same datetime object passed in
    """
    response =plc.read_plc_time(dt=datetime.now())



    """
    Synchronize PLC time to PC time
        example: Set PLC time using UTC time

    Args:
        utc(bool): True to use UTC time and false using local time

    Returns:
        dt(datetime): the datetime object that was used
            example:
                2022-09-08 17:07:25.380000
    """
    response = plc.sync_plc_time(utc=True)



    """
    Read CPU Status

    Returns:
        CPUStatus(NamedTuple): (requestedMode(str), previousMode(str), error(str))
            example:
                Run,Unknown,
    """
    response = plc.read_cpu_status()



    """
    Read catalog (order) code

    Returns:
        CatalogCode(NamedTuple): (
            moduleOrderNumber(str), 
            moduleVersion(str), 
            basicHardwareId(str),
            hardwareVersion(str),
            basicFirmwareId(str),
            firmwareVersion(str),
            firmwareExtensionId(str),
            firmwareExtVersion(str),
            error(str)
        )
            example: 
                6ES7 318-3FL01-0AB0,9.1,6ES7 318-3FL01-0AB0,9.1,,22019.525,Boot Loader,16677.3084,
    """
    response = plc.read_catalog_code()



    """
    Read CPU information

    Returns:
        CatalogCode(NamedTuple): (
            systemName(str), 
            moduleName(str), 
            plantId(str),
            copyright(str),
            serialNumber(str),
            cpuType(str),
            memSerialNumber(str),
            manufacturerId(str),
            profileId(str),
            profileSpec(str),
            oemCopyright(str),
            oemId(str),
            oemAddId(str),
            locationId(str),
            error(str)
        )
            example:
                S7300/ET200M station_1,PLC_1,,Original Siemens Equipment,
                S C-J3LX93592017,CPU 319F-3 PN/DP,MMC 86A97F09,0x002a,
                0xf600,0x0001,,0x0000,0x00000000,,
    """
    response = plc.read_cpu_info()



    """
    Read communication processor

    Returns:
        list[CommProc]:
            example:
                [
                    CommProc(maxPDU=240, maxConnections=32, mpiRate=12000000, mkbusRate=187500, error='')
                ]
    """
    response = plc.read_comm_proc()



    """
    Stop PLC

    Returns:
        bool: True for successful, and False for failed
    """
    response = plc.stop()



    """
    Start PLC with initial settings

    Returns:
        bool: True for successful, and False for failed
    """
    response = plc.start_plc_cold()



    """
    Start PLC from previous state

    Returns:
        bool: True for successful, and False for failed
    """
    response = plc.start_plc_hot()



    """
    Read system state list (SZL-ID)

    Returns:
        bytes: Raw bytes that would need parsing
    """
    response = plc.read_szl(id=0x001C, index=0x0000)



    """
    Read protection level

    Returns:
        list[Protection]: (
            protectionLevel(int),
            passwordLevel(int),
            validProtectionLevel(int),
            modeSelector(str),
            startupSwitch(str),
            error(str)
        )
            example:
                [
                    Protection(
                        protectionLevel=1, 
                        passwordLevel=0, 
                        validProtectionLevel=1, 
                        modeSelector='Run Program', 
                        startupSwitch='Undefined', 
                        error=''
                    )
                ]
    """
    response = plc.read_protection()



    """
    Read CPU diagnostics

    Returns:
        list[CPUDiagnostics]: (
            eventId(str),
            description(str),
            priority(int),
            obNumber(int),
            datId(str),
            info1(str),
            info2(str),
            timestamp(datetime),
            error(str)
        )
            example:
                [
                    CPUDiagnostics(
                        eventId='0x4302', 
                        description='Mode transition from STARTUP to RUN', 
                        priority=255, 
                        obNumber=104, 
                        datId='0xc700', 
                        info1='0x0000', 
                        info2='0x08147714', 
                        timestamp=datetime.datetime(2022, 9, 8, 15, 23, 19, 521000), 
                        error=''
                    ),
                    ... 
                    CPUDiagnostics(
                        eventId='0x4302', 
                        description='Mode transition from STARTUP to RUN', 
                        priority=255, 
                        obNumber=104, 
                        datId='0xc700', 
                        info1='0x0000', 
                        info2='0x08147714', 
                        timestamp=datetime.datetime(2022, 9, 8, 8, 59, 44, 546000), 
                        error=''
                    )
                ]
    """
    response = plc.read_protection()



    """
    Read CPU LEDs

    Returns:
        list[CPULed]: (
            rack(int),
            type(int),
            id(str),
            on(int),
            flashing(bool),
            error(str)
        )
            example:
                [
                    CPULed(
                        rack=0, 
                        type=0, 
                        id='SF (group error)', 
                        on=False, 
                        flashing=False, 
                        error=''
                    ),
                    ...
                    CPULed(
                        rack=0, 
                        type=0, 
                        id='MAINT (maintenance demand)', 
                        on=False, 
                        flashing=False, 
                        error=''
                    )
                ]
    """
    response = plc.read_cpu_leds()



    """
    Read block info
        example: read info of data block 1

    Args:
        BlockType(int):
        BlockNumber(int):

    Returns:
        BlockInfo(NamedTuple): (
            flags(str),
            language(str),
            type(str),
            number(int),
            loadMemory(int),
            security(str),
            codeTimestamp(str),
            interfaceTimestamp(str),
            ssbLength(int),
            addLength(int),
            localDataLength(int),
            mc7Length(int),
            author(str),
            family(str),
            name(str),
            version(str),
            checksum(str),
            error(str)
        )
            example:
                [
                    BlockInfo(
                        flags='0x01', 
                        language='DB', 
                        type='DB', 
                        number=1, 
                        loadMemory=410, 
                        security=0, 
                        codeTimestamp=datetime.datetime(2022, 9, 8, 5, 44, 42, 465000), 
                        interfaceTimestamp=datetime.datetime(2022, 9, 6, 16, 48, 45, 568000), 
                        ssbLength=42, 
                        addLength=0, 
                        localDataLength=0, 
                        mc7Length=296, 
                        author='', 
                        family='', 
                        name='', 
                        version='10.1', 
                        checksum='0xf3db', 
                        error=''
                    )
                ]
    """
    response = plc.read_block_info(BlockType=BlockType.DB, BlockNumber=1)



    """
    Read area
        example: read area memory of data block 1 with no offsets using a list of items

    Args:
        ItemList(list)[Required]
    Notes:
        Only data block uses block number and the rest is set to 0
            Example: I10.1 is not valid and will be override to I0.1
                     Q1.1 is not valid and will be override to Q0.1
                     DB20.DBX1.0 is valid and produces block number 10
                        with byte offset of 1 and bit offset of 0

    Returns:
        list(Tag): list that contains parsed data
            example:
                [
                    Tag(name="CHAR", address="DB2.DBX2.0", value=b'a', size=2, type=3, error=''),
                    Tag(name="BYTE", address="DB2.DBX1.0", value=1, size=1, type=2, error=''),
                    Tag(name="DINT", address="DB2.DBX4.0", value=-10000, size=4, type=7, error=''),
                    Tag(name="DWORD", address="DB2.DBX8.0", value=10000, size=4, type=6, error=''),
                    ...
                    Tag(name="BOOL4", address="DB2.DBX0.4", value=False, size=1, type=1, error=''),
                    Tag(name="FLAG", address="M0.4", value=False, size=1, type=1, error=''),
                    Tag(name="INPUT 3", address="I0.3", value=True, size=1, type=1, error=''),
                    Tag(name="COUNTER0", address="C0", value=0, size=1, type=28, error=''),
                    Tag(name="TIMER0", address="T0", value=0, size=1, type=29, error=''),
                ]
    """
    response = plc.read_area(item_list=__RTAGS1)



    """
    Write area
        example: write area memory of data block 2 with no offsets using a list of items

    Args:
        ItemList(list)[Required]
    Notes:
        Only data block uses block number and the rest is set to 0
            Example: I10.1 is not valid and will be override to I0.1
                     Q1.1 is not valid and will be override to Q0.1
                     DB20.1 is valid and produces block number 10 with offset of 1
    Returns:
        list(Tag): same input list with error update
            example:
                [
                    Tag(name='BIT', address='DB2.DBX0.0', size=1, type=1, error='Missing Value'), 
                    Tag(name='BYTE', address='DB2.DBX1.0', value=1, size=1, type=2, error=''), 
                    Tag(name='CHAR', address='DB2.DBX2.0', value=b'a', size=2, type=3, error=''), 
                    Tag(name='DINT', address='DB2.DBX4.0', value=1, size=4, type=7, error=''), 
                    Tag(name='DWORD', address='DB2.DBX8.0', value=1, size=4, type=6, error=''), 
                    ... 
                    Tag(name='WORD', address='DB2.DBX500.0', value=1, size=2, type=4, error='Invalid address')
                ]

    """
    plc.write_area(item_list=__WTAGS2)
```


CHANGELOG
v0.2.1:
* handle invalid address calls

v0.2.0:
* big refactor {read,write}_area to match controller syntax
* add iec_{counter,timer}
* move encode/decode into utility

v0.1.2:
* update {read/write}_area with controller type since readblock doesn't work for S71500

v0.1.1:
* simplify read/write by merging Area, Number, Offset into Address

v0.1.0:
* Initial commit


pystep7/tag.py
"""
This file is a collection of data storage mechanisms.
"""

from typing import NamedTuple, Any, Optional
from reprlib import repr as _r


__all__ = ["Tag"]


class Tag(NamedTuple):
    name: Optional[str] = ''        #: friendly ID
    address: Optional[str] = ''     #: address space
    value: Optional[Any] = None     #: value read/written, may be ``None`` on error
    size: Optional[int] = 0         #: length of value, useful for string
    type: Optional[str] = ''        #: data type of element (bit, byte, char, etc)
    error: Optional[str] = ''       #: error message if unsuccessful, else ``None``


    def __bool__(self):
        """
        ``True`` if both ``value`` is not ``None`` and ``error`` is ``None``
        ``False`` otherwise
        """
        return self.value is not None and self.error is None


    def __str__(self):
        return (
            f"{self.name}, "
            f"{self.address}, "
            f"{_r(self.value)}, "
            f"{self.size}, "
            f"{self.type}, "
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"name={self.name!r}, "
            f"address={self.address!r}, "
            f"value={self.value!r}, "
            f"size={self.size!r}, "
            f"type={self.type!r}, "
            f"error={self.error!r})"
        )


class CPUInfo(NamedTuple):
    systemName: Optional[str] = ''      # name (e.g. 'R08ENCPU')
    moduleName: Optional[str] = ''      # code (e.g. '4806')
    plantId: Optional[str] = ''         # code (e.g. '4806')
    copyright: Optional[str] = ''       # code (e.g. '4806')
    serialNumber: Optional[str] = ''    # code (e.g. '4806')
    cpuType: Optional[str] = ''         # code (e.g. '4806')
    memSerialNumber: Optional[str] = '' # code (e.g. '4806')
    manufacturerId: Optional[str] = ''  # code (e.g. '4806')
    profileId: Optional[str] = ''       # code (e.g. '4806')
    profileSpec: Optional[str] = ''     # code (e.g. '4806')
    oemCopyright: Optional[str] = ''    # code (e.g. '4806')
    oemId: Optional[str] = ''           # code (e.g. '4806')
    oemAddId: Optional[str] = ''        # code (e.g. '4806')
    locationId: Optional[str] = ''      # code (e.g. '4806')
    error: Optional[str] = ''


    def __str__(self):
        return (
            f"{self.systemName},"
            f"{self.moduleName},"
            f"{self.plantId},"
            f"{self.copyright},"
            f"{self.serialNumber},"
            f"{self.cpuType},"
            f"{self.memSerialNumber},"
            f"{self.manufacturerId},"
            f"{self.profileId},"
            f"{self.profileSpec},"
            f"{self.oemCopyright},"
            f"{self.oemId},"
            f"{self.oemAddId},"
            f"{self.locationId},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"systemName={self.systemName!r}, "
            f"moduleName={self.moduleName!r}, "
            f"plantId={self.plantId!r}, "
            f"copyright={self.copyright!r}, "
            f"serialNumber={self.serialNumber!r}, "
            f"cpuType={self.cpuType!r}, "
            f"memSerialNumber={self.memSerialNumber!r}, "
            f"manufacturerId={self.manufacturerId!r}, "
            f"profileId={self.profileId!r}, "
            f"profileSpec={self.profileSpec!r}, "
            f"oemCopyright={self.oemCopyright!r}, "
            f"oemId={self.oemId!r}, "
            f"oemAddId={self.oemAddId!r}, "
            f"locationId={self.locationId!r}"
            f"error={self.error!r}"
            ")"
        )


class CatalogCode(NamedTuple):
    moduleOrderNumber: Optional[str] = ''   # status (e.g. 'Stop')
    moduleVersion: Optional[str] = ''
    basicHardwareId: Optional[str] = ''     # status (e.g. 'Unknown')
    hardwareVersion: Optional[str] = ''
    basicFirmwareId: Optional[str] = ''     # status (e.g. 'Unknown')
    firmwareVersion: Optional[str] = ''
    firmwareExtensionId: Optional[str] = '' # status (e.g. 'Unknown')
    firmwareExtVersion: Optional[str] = ''  # status (e.g. '9.1')
    error: Optional[str] = ''


    def __str__(self):
        return (
            f"{self.moduleOrderNumber},"
            f"{self.moduleVersion},"
            f"{self.basicHardwareId},"
            f"{self.hardwareVersion},"
            f"{self.basicFirmwareId},"
            f"{self.firmwareVersion},"
            f"{self.firmwareExtensionId},"
            f"{self.firmwareExtVersion},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"moduleOrderNumber={self.moduleOrderNumber!r}, "
            f"moduleVersion={self.moduleVersion!r}, "
            f"basicHardwareId={self.basicHardwareId!r}), "
            f"hardwareVersion={self.hardwareVersion!r}), "
            f"basicFirmwareId={self.basicFirmwareId!r}, "
            f"firmwareVersion={self.firmwareVersion!r}, "
            f"firmwareExtensionId={self.firmwareExtensionId!r}, "
            f"firmwareExtVersion={self.firmwareExtVersion!r}, "
            f"error={self.error!r}"
            ")"
        )


class CPUStatus(NamedTuple):
    requestedMode: Optional[str] = ''   # status (e.g. 'Stop')
    previousMode: Optional[str] = ''    # status (e.g. 'Unknown')
    error: Optional[str] = ''

    def __str__(self):
        return f"{self.requestedMode},{self.previousMode},{self.error}"


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"requestedMode={self.requestedMode!r}, "
            f"previousMode={self.previousMode!r}, "
            f"error={self.error!r}"
            ")"
        )


class CommProc(NamedTuple):
    maxPDU: Optional[int] = 0           # status (e.g. 240)
    maxConnections: Optional[int] = 0   # status (e.g. 32)
    mpiRate: Optional[int] = 0          # status (e.g. 12000000)
    mkbusRate: Optional[int] = 0        # status (e.g. 187500)
    error: Optional[str] = ''


    def __str__(self):
        return (
            f"{self.maxPDU},"
            f"{self.maxConnections},"
            f"{self.mpiRate},"
            f"{self.mkbusRate},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"maxPDU={self.maxPDU!r}, "
            f"maxConnections={self.maxConnections!r}, "
            f"mpiRate={self.mpiRate!r}, "
            f"mkbusRate={self.mkbusRate!r}, "
            f"error={self.error!r}"
            ")"
        )


class Protection(NamedTuple):
    protectionLevel: Optional[int] = 1      # protection level (1,2,3)
    passwordLevel: Optional[int] = 0        # password level (0,1,2,3)
    validProtectionLevel: Optional[int] = 0 # valid protection level (0,1,2,3)
    modeSelector: Optional[str] = ''        # mode selector setting
                                            #   1: Run
                                            #   2: Run-Protected
                                            #   3: Stop
                                            #   4: Memory Reset
    startupSwitch: Optional[str] = ''       # startup switch setting (0,1,2)
                                            #   0: Undefined
                                            #   1: Cold Restart
                                            #   2: Warm Restart
    error: Optional[str] = ''


    def __str__(self):
        return (
            f"{self.protectionLevel},"
            f"{self.passwordLevel},"
            f"{self.validProtectionLevel},"
            f"{self.modeSelector},"
            f"{self.startupSwitch},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"protectionLevel={self.protectionLevel!r}, "
            f"passwordLevel={self.passwordLevel!r}, "
            f"validProtectionLevel={self.validProtectionLevel!r}, "
            f"modeSelector={self.modeSelector!r}, "
            f"startupSwitch={self.startupSwitch!r}, "
            f"error={self.error!r}"
            ")"
        )


class BlockInfo(NamedTuple):
    flags: Optional[str] = ''
    language: Optional[str] = ''
    type: Optional[str] = ''
    number: Optional[int] = 0
    loadMemory: Optional[int] = 0
    security: Optional[str] = ''
    codeTimestamp: Optional[str] = ''
    interfaceTimestamp: Optional[str] = ''
    ssbLength: Optional[int] = 0
    addLength: Optional[int] = 0
    localDataLength: Optional[int] = 0
    mc7Length: Optional[int] = 0
    author: Optional[str] = ''
    family: Optional[str] = ''
    name: Optional[str] = ''
    version: Optional[str] = ''
    checksum: Optional[str] = ''
    error: Optional[str] = ''

    def __str__(self):
        return (
            f"{self.flags},"
            f"{self.language},"
            f"{self.type},"
            f"{self.number},"
            f"{self.loadMemory},"
            f"{self.security},"
            f"{self.codeTimestamp},"
            f"{self.interfaceTimestamp},"
            f"{self.ssbLength},"
            f"{self.addLength},"
            f"{self.localDataLength},"
            f"{self.mc7Length},"
            f"{self.author},"
            f"{self.family},"
            f"{self.name},"
            f"{self.version},"
            f"{self.checksum},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"flags={self.flags!r}, "
            f"language={self.language!r}, "
            f"type={self.type!r}, "
            f"number={self.number!r}, "
            f"loadMemory={self.loadMemory!r}, "
            f"security={self.security!r}, "
            f"codeTimestamp={self.codeTimestamp!r}, "
            f"interfaceTimestamp={self.interfaceTimestamp!r}, "
            f"ssbLength={self.ssbLength!r}, "
            f"addLength={self.addLength!r}, "
            f"localDataLength={self.localDataLength!r}, "
            f"mc7Length={self.mc7Length!r}, "
            f"author={self.author!r}, "
            f"family={self.family!r}, "
            f"name={self.name!r}, "
            f"version={self.version!r}, "
            f"checksum={self.checksum!r}, "
            f"error={self.error!r}"
            ")"
        )


class CPUDiagnostics(NamedTuple):
    eventId: Optional[str] = ''     #
    description: Optional[str] = '' #
    priority: Optional[int] = 0     #
    obNumber: Optional[int] = 0     #
    datId: Optional[str] = ''       #
    info1: Optional[str] = ''       #
    info2: Optional[str] = ''       #
    timestamp: Optional[Any] = None #
    error: Optional[str] = ''       #


    def __str__(self):
        return (
            f"{self.eventId},"
            f"{self.description},"
            f"{self.priority},"
            f"{self.obNumber},"
            f"{self.datId},"
            f"{self.info1},"
            f"{self.info2},"
            f"{self.timestamp},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"eventId={self.eventId!r}, "
            f"description={self.description!r}, "
            f"priority={self.priority!r}, "
            f"obNumber={self.obNumber!r}, "
            f"datId={self.datId!r}, "
            f"info1={self.info1!r}, "
            f"info2={self.info2!r}, "
            f"timestamp={self.timestamp!r}, "
            f"error={self.error!r}"
            ")"
        )


class CPULed(NamedTuple):
    rack: Optional[int] = 0             #
    type: Optional[int] = 0             #
    id: Optional[str] = ''              #
    on: Optional[int] = 0               #
    flashing: Optional[bool] = False    #
    error: Optional[str] = ''           #

    def __str__(self):
        return (
            f"{self.rack},"
            f"{self.type},"
            f"{self.id},"
            f"{self.on},"
            f"{self.flashing},"
            f"{self.error}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"rack={self.rack!r}, "
            f"type={self.type!r}, "
            f"id={self.id!r}, "
            f"on={self.on!r}, "
            f"flashing={self.flashing!r}, "
            f"error={self.error!r}"
            ")"
        )


class IecCounter(NamedTuple):
    """
    Input
        C_DU(bool):     Counter type D(down) U(up) [offset 0.0]
        LOAD_R(bool):   LOAD (down) R (up) [offset 0.1]
        PV(int):        Preset value [offset 2.0]
    Output
        Q(bool):        Done flag [offset 4.0]
        CV(int):        Current value [offset 6.0]
    InOut
        C_DU_O(bool):   Counter D(down) U(up) [offset 8.0]
    """
    C_DU: Optional[bool] = False
    LOAD_R: Optional[bool] = False
    PV: Optional[int] = 0x0000
    Q: Optional[bool] = False
    CV: Optional[int] = 0x0000
    C_DU_O: Optional[bool] = False

    def __str__(self):
        return (
            f"{self.C_DU},"
            f"{self.LOAD_R},"
            f"{self.PV},"
            f"{self.Q},"
            f"{self.CV},"
            f"{self.C_DU_O}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"C_DU={self.C_DU!r}, "
            f"LOAD_R={self.LOAD_R!r}, "
            f"PV={self.PV!r}, "
            f"Q={self.Q!r}, "
            f"CV={self.CV!r}, "
            f"C_DU_O={self.C_DU_O!r}"
            ")"
        )


class IecTimer(NamedTuple):
    """
    Input
        IN(bool):       Enable flag [offset 0.0]
        PT(int):        Preset time (TIME) [offset 2.0]
    Output
        Q(bool):        Done flag [offset 6.0]
        ET(int):        Elapsed time (TIME) [offset 8.0]
    InOut
        STATE(int):     (BYTE) [offset 12.0]
        STIME(int):     (TIME) [offset 14.0]
        ATIME(int):     (TIME) [offset 18.0]
    """
    IN: Optional[bool] = False
    PT: Optional[int] = 0x0000
    Q: Optional[bool] = False
    ET: Optional[int] = 0x0000
    STATE: Optional[int] = 0x00
    STIME: Optional[int] = 0x0000
    ATIME: Optional[int] = 0x0000

    def __str__(self):
        return (
            f"{self.IN},"
            f"{self.PT},"
            f"{self.Q},"
            f"{self.ET},"
            f"{self.STATE},"
            f"{self.STIME},"
            f"{self.ATIME}"
        )


    def __repr__(self):
        return (
            f"{self.__class__.__name__}("
            f"IN={self.IN!r}, "
            f"PT={self.PT!r}, "
            f"Q={self.Q!r}, "
            f"ET={self.ET!r}, "
            f"STATE={self.STATE!r}, "
            f"STIME={self.STIME!r}, "
            f"ATIME={self.ATIME!r}"
            ")"
        )
